File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
           PriorityQueue<ListNode> pq=new PriorityQueue<>((a,b)->a.val-b.val);
        for (ListNode list : lists) {
            if(list!=null) pq.add(list);
        }
        ListNode tmp=new ListNode(-1);
        ListNode cur=tmp;
        while(!pq.isEmpty()){
            ListNode poll = pq.poll();
            cur.next=poll;
            cur=cur.next;
            if(poll.next!=null) pq.add(poll.next);
        }
        return tmp.next;
    }
}

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        minHeap = []
        for i, node in enumerate(lists):
            if node:
                heapq.heappush(minHeap, (node.val, i, node))
        dummy = ListNode()
        cur = dummy
        while minHeap:
            _, i, node = heapq.heappop(minHeap)
            cur.next = node
            cur = cur.next
            if node.next:
                heapq.heappush(minHeap, (node.next.val, i, node.next))
        return dummy.next
        